group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,446 @@
+                            "to maintain the requested number of standby replicas.");
+                    break;
+                }
+                Owner standby = null;
+
+                // prev active task
+                Owner prevOwner = activeTaskToPrevOwner.get(task);
+                if (prevOwner != null && availableProcesses.contains(prevOwner.processId) && isLoadBalanced(prevOwner.processId)
+                        && taskPairs.hasNewPair(task, processIdToState.get(prevOwner.processId).assignedTasks())) {
+                    standby = prevOwner;
+                }
+
+                // prev standby tasks
+                if (standby == null) {
+                    final Set<Owner> prevOwners = standbyTaskToPrevOwner.get(task);
+                    if (prevOwners != null && !prevOwners.isEmpty()) {
+                        prevOwners.removeIf(owner -> !availableProcesses.contains(owner.processId));
+                        prevOwner = findOwnerWithLeastLoad(prevOwners, task, true);
+                        if (prevOwner != null && isLoadBalanced(prevOwner.processId)) {
+                            standby = prevOwner;
+                        }
+                    }
+                }
+
+                // others
+                if (standby == null) {
+                    standby = findOwnerWithLeastLoad(task, availableProcesses);
+                }
+
+                standbyTasksAssignments.putIfAbsent(standby.memberId, new HashSet<>());
+                standbyTasksAssignments.get(standby.memberId).add(task);
+                updateHelpers(standby, task, false);
+            }
+
+        }
+    }
+
+    private boolean isLoadBalanced(final String processId) {
+        final ProcessState process = processIdToState.get(processId);
+        return process.hasCapacity() || isLeastLoadedProcess(process.load());
+    }
+
+    private boolean isLeastLoadedProcess(final double load) {
+        return processIdToState.values().stream()
+                .allMatch(process -> process.load() >= load);
+    }
+
+    private Set<String> findAllowedProcesses(final TaskId taskId) {
+        return processIdToState.values().stream()
+                .filter(process -> !process.hasTask(taskId))
+                .map(ProcessState::processId)
+                .collect(Collectors.toSet());
+    }
+
+    private void updateHelpers(final Owner owner, final TaskId taskId, final boolean isActive) {
+        // add all pair combinations: update taskPairs
+        taskPairs.addPairs(taskId, processIdToState.get(owner.processId).assignedTasks());
+
+        ProcessState process = processIdToState.get(owner.processId);
+        process.addTask(owner.memberId, taskId, isActive);
+
+        if (isActive) {
+            // update task per process
+            maybeUpdateTasksPerProcess(process.assignedActiveTasks().size());
+        }
+    }
+
+    private void computeTasksPerProcess() {
```

Comment:
This seems to be a computation that might be specified as static. If it is not necessary, I would not work with the state of the object. This is like using a global variable, thus error-prone.
I would specify the method as
 ```suggestion
    private static int computeTasksPerProcess(final int numberOfTasks, final int numberOfMembers) {
```